### PR TITLE
Move websocket code to common. Add OkHttp & Js websockets

### PR DESCRIPTION
**Pull request by e5l**

### CODE REVIEW

File: ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/FrameJvm.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import io.ktor.util.*
+import kotlinx.coroutines.*
+import java.nio.*
+
+/**
+ * Frame content
+ */
+val Frame.buffer: ByteBuffer
+    get() = ByteBuffer.wrap(data)
```
Comment (cy6erGn0m) [inline ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/FrameJvm.kt]: This is breaking change: the problem is that it doesn't keep state between invocations

File: ktor-utils/common/src/io/ktor/util/Bytes.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package io.ktor.util
+
+/**
+ * Read [Short] with specified [offset] from [ByteArray].
```
Comment (cy6erGn0m) [inline ktor-utils/common/src/io/ktor/util/Bytes.kt]: Notice endian in kdoc

File: ktor-utils/js/src/io/ktor/util/CryptoJs.kt [hunk 3]
```diff
@@ -40,3 +40,11 @@ private external object crypto {
 private external class SubtleCrypto {
     fun digest(algoName: String, buffer: ByteArray): Promise<ArrayBuffer>
 }
+
+/**
+ * Compute SHA-1 hash for the specified [bytes]
+ */
+@KtorExperimentalAPI
+actual fun sha1(bytes: ByteArray): ByteArray {
+    TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
```
Comment (cy6erGn0m) [inline ktor-utils/js/src/io/ktor/util/CryptoJs.kt]: error message

File: ktor-utils/posix/src/io/ktor/util/CryptoNative.kt [hunk 4]
```diff
@@ -5,3 +5,11 @@ actual fun generateNonce(): String = error("[generateNonce] is not supported on
 
 @InternalAPI
 actual fun Digest(name: String): Digest = error("[Digest] is not supported on iOS")
+
+/**
+ * Compute SHA-1 hash for the specified [bytes]
+ */
+@KtorExperimentalAPI
+actual fun sha1(bytes: ByteArray): ByteArray {
+    TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
```
Comment (cy6erGn0m) [inline ktor-utils/posix/src/io/ktor/util/CryptoNative.kt]: Same: message and the generated comment

File: ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/iosArm32/src/TestIosArm32.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+actual fun testSuspend(
+    context: CoroutineContext,
+    block: suspend CoroutineScope.() -> Unit
+): Unit = runBlocking {
+    val loop = coroutineContext[ContinuationInterceptor] as EventLoop
+
+    val task = launch { block() }
+    while (!task.isCompleted) {
+        val date = NSDate().addTimeInterval(1.0) as NSDate
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/iosArm32/src/TestIosArm32.kt]: `addTimeInterval` is deprecated: https://developer.apple.com/documentation/foundation/nsdate/1591575-addtimeinterval

File: ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/iosArm32/src/TestIosArm32.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        NSRunLoop.mainRunLoop.runUntilDate(date)
+
+        loop.processNextEvent()
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/iosArm32/src/TestIosArm32.kt]: Shouldn't we use the returned value somehow? E.g. we can use it to evaluate `date`

File: ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/js/src/TestJs.kt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+package io.ktor.client.tests.utils.dispatcher
+
+import kotlinx.coroutines.*
+import kotlin.coroutines.*
+
+/**
+ * Test runner for common suspend tests.
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/js/src/TestJs.kt]: > ... common ...
not exactly precise

File: ktor-features/ktor-websockets/jvm/test/io/ktor/tests/websocket/WebSocketEngineSuite.kt [hunk 8]
```diff
@@ -105,16 +109,19 @@ abstract class WebSocketEngineSuite<TEngine : ApplicationEngine, TConfiguration
                     incoming.consumeEach {
                     }
                 } catch (cancelled: CancellationException) {
+                    println("$cancelled: here")
```
Comment (cy6erGn0m) [inline ktor-features/ktor-websockets/jvm/test/io/ktor/tests/websocket/WebSocketEngineSuite.kt]: :printer:

File: ktor-features/ktor-websockets/jvm/test/io/ktor/tests/websocket/WebSocketEngineSuite.kt [hunk 9]
```diff
@@ -105,16 +109,19 @@ abstract class WebSocketEngineSuite<TEngine : ApplicationEngine, TConfiguration
...
                 } catch (t: Throwable) {
                     errors.addError(t)
+                    println("$t is here")
```
Comment (cy6erGn0m) [inline ktor-features/ktor-websockets/jvm/test/io/ktor/tests/websocket/WebSocketEngineSuite.kt]: :printer:

File: ktor-http/common/src/io/ktor/http/HttpProtocolVersion.kt [hunk 10]
```diff
@@ -36,13 +36,26 @@ data class HttpProtocolVersion(val name: String, val major: Int, val minor: Int)
...
+         * Creates an instance of [HttpProtocolVersion] from the given parameters.
          */
         fun fromValue(name: String, major: Int, minor: Int): HttpProtocolVersion = when {
             name == "HTTP" && major == 1 && minor == 1 -> HTTP_1_1
             name == "HTTP" && major == 2 && minor == 0 -> HTTP_2_0
             else -> HttpProtocolVersion(name, major, minor)
         }
+
+        /**
+         * Create an instance of [HttpProtocolVersion] from htto string representation.
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/HttpProtocolVersion.kt]:
```suggestion
         * Create an instance of [HttpProtocolVersion] from http string representation.
```

File: ktor-http/common/src/io/ktor/http/HttpProtocolVersion.kt [hunk 11]
```diff
@@ -36,13 +36,26 @@ data class HttpProtocolVersion(val name: String, val major: Int, val minor: Int)
...
+         */
+        fun parse(value: CharSequence): HttpProtocolVersion {
+            /**
+             * protocol/major.minor
+             */
+            val (protocol, major, minor) = value.split("/", ".")
+                .also { check(it.size == 3) }
```
Comment (cy6erGn0m) [inline ktor-http/common/src/io/ktor/http/HttpProtocolVersion.kt]: `check` failure message is required here to explain what requirement failed

File: ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/Frame.kt [hunk 12]
```diff
@@ -78,98 +44,100 @@ sealed class Frame(val fin: Boolean, val frameType: FrameType, val buffer: ByteB
...
  */
 fun Frame.Close.readReason(): CloseReason? {
-    if (buffer.remaining() < 2) {
+    if (data.size < 2) {
         return null
     }
 
     buffer.mark()
-    val code = buffer.getShort()
+    val code = buffer.short
```
Comment (cy6erGn0m) [inline ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/Frame.kt]: `buffer.short` looks like there is some kind short buffer representation or a buffer or short integers rather that reading a short integer number. This is why I usually don't like to apply the intension.

File: ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/RawWebSocket.kt [hunk 13]
```diff
@@ -47,7 +47,7 @@ class RawWebSocket(
     }
 
     override suspend fun close(cause: Throwable?) {
-        terminate()
+        cause?.let { socketJob.completeExceptionally(it) } ?: socketJob.complete(Unit)
```
Comment (cy6erGn0m) [inline ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/RawWebSocket.kt]: `socketJob` successful completion doesn't close anything, isn't it? Doesn't closing without exception lead to forever hanging raw read/write coroutines?

File: ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/WebSocketSessionJvm.kt [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    actual val outgoing: SendChannel<Frame>
+
+    /**
+     * Enqueue frame, may suspend if outgoing queue is full. May throw an exception if outgoing channel is already
+     * closed so it is impossible to transfer any message. Frames that were sent after close frame could be silently
+     * ignored. Please note that close frame could be sent automatically in reply to a peer close frame unless it is
+     * raw websocket session.
+     */
+    @Suppress("ACTUAL_WITHOUT_EXPECT")
```
Comment (cy6erGn0m) [inline ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/WebSocketSessionJvm.kt]: Do we need this suppression? Why?

File: ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/WebSocketSessionJvm.kt [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * Initiate connection termination immediately. Termination may complete asynchronously.
+     */
+    actual fun terminate()
+
+    /**
+     * Close session with the specified [cause] or with no reason if `null`
+     */
+    @Suppress("ACTUAL_FUNCTION_WITH_DEFAULT_ARGUMENTS")
+    actual suspend fun close(cause: Throwable? = null)
```
Comment (cy6erGn0m) [inline ktor-http/ktor-http-cio/jvm/src/io/ktor/http/cio/websocket/WebSocketSessionJvm.kt]: Shouldn't the default argument be specified on the expect-declaration?

File: gradle/posix.gradle [hunk 16]
```diff
@@ -33,9 +33,18 @@ kotlin {
...
                 dependsOn posixTest
             }
+
+            iosArm32Test.dependsOn iosArm32Main
+            iosArm64Test.dependsOn iosArm64Main
+            iosX64Test.dependsOn iosX64Main
+            linuxX64Test.dependsOn linuxX64Main
+            macosX64Test.dependsOn macosX64Main
+            iosX64Test.dependsOn iosX64Main
+            mingwX64Test.dependsOn mingwX64Main
```
Comment (cy6erGn0m) [inline gradle/posix.gradle]: :angry:

File: ktor-client/ktor-client-features/ktor-client-websocket/js/src/io/ktor/client/features/websocket/JsWebSocketSession.kt [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+
+    override fun terminate() {
+        _incoming.cancel()
+        _outgoing.cancel()
+        websocket.close()
+    }
+
+    @KtorExperimentalAPI
+    override suspend fun close(cause: Throwable?) {
+        _incoming.send(Frame.Close())
```
Comment (cy6erGn0m) [inline ktor-client/ktor-client-features/ktor-client-websocket/js/src/io/ktor/client/features/websocket/JsWebSocketSession.kt]: Should't we send some specific close reason if `cause != null` ?

File: ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/iosArm32/src/TestIosArm32.kt (hunk 5 above)
Comment (e5l) [reply ktor-client/ktor-client-tests/ktor-client-tests-dispatcher/iosArm32/src/TestIosArm32.kt]: There is no `addingTimeInterval` method in `kotlin-native` headers and it doesn't compile.

**Pull request merged**